File: lib/selenium.ts
import { Browser, Builder } from "selenium-webdriver";
import chrome from "selenium-webdriver/chrome";
import { solveRecaptchaV2 } from "./2captcha";
import { By, until } from "selenium-webdriver";

const service = new chrome.ServiceBuilder(
  process.cwd() + "/assets/selenium/chromedriver"
);

export async function getGoogleAutocompleteSuggestions(
  keyword: string,
  uule: string
) {
  console.log(process.cwd() + "/assets/selenium/chromedriver");
  let driver = await new Builder()
    .forBrowser(Browser.CHROME)
    .setChromeService(service)
    .build();
  try {
    const pageUrl = `https://www.google.com/search?q=${keyword}&uule=${uule}`;

    await driver.get(pageUrl);

    const hasCaptchaForm = await driver.findElement(By.id("captcha-form"));

    console.log(!!hasCaptchaForm, "hasCaptchaForm");

    if (hasCaptchaForm) {
      const googleSiteKey = await driver
        .findElement(By.css("[data-sitekey]"))
        .getAttribute("data-sitekey");

      const data_s = await driver
        .findElement(By.css(".g-recaptcha"))
        .getAttribute("data-s");

      console.log(googleSiteKey, "googleSiteKey");
      console.log(data_s, "data_s");

      const result = await solveRecaptchaV2(pageUrl, googleSiteKey, data_s);

      const token = result.data.data;

      if (token) {
        await driver.executeScript(
          `
            document.getElementById('g-recaptcha-response').innerHTML = arguments[0];
            document.getElementById('g-recaptcha-response').style.display = 'block';
            `,
          token
        );

        await driver.executeScript(
          `document.getElementById('captcha-form').submit();`
        );

        await driver.wait(until.stalenessOf(hasCaptchaForm), 10000);
      }
    }

    await new Promise((resolve) => setTimeout(resolve, 50000000));
  } catch (error) {
    console.error(error);
    return { error };
  } finally {
    await driver.quit();
  }
}

File: api/lib/selenium.py
from lib.openvpn import openvpn_connection
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from lib.twocaptcha import solve_recaptcha2


def get_keywords(keyword, uule, openvpn_config):
    print("get_keywords", keyword, uule, openvpn_config)

    driver = None
    try:
        keywords = {
            "autocomplete": [],
        }
        with openvpn_connection(openvpn_config):
            driver = webdriver.Chrome()
            driver.get(f"https://www.google.com/search?q={keyword}&uule={uule}")

            captcha_form = driver.find_element(By.ID, "captcha-form")
            print("captcha_form", bool(captcha_form))
            if captcha_form:
                sitekey = captcha_form.find_element(
                    By.CSS_SELECTOR, "[data-sitekey]"
                ).get_attribute("data-sitekey")
                print("sitekey", sitekey)

                data_s = captcha_form.find_element(
                    By.CSS_SELECTOR, ".g-recaptcha"
                ).get_attribute("data-s")
                print("data_s", data_s)

                captcha_result = solve_recaptcha2(
                    "https://www.google.com/recaptcha/api2/demo",
                    sitekey,
                    data_s
                )
                print("captcha_result", captcha_result)
                captcha_token = captcha_result["code"]
                print("captcha_token", captcha_token)

                driver.execute(
                    '''
                        document.getElementById('g-recaptcha-response').innerHTML = arguments[0];
                        document.getElementById('g-recaptcha-response').style.display = 'block';
                        document.getElementById('captcha-form').submit();
                    ''',
                    captcha_token
                )

                WebDriverWait(driver, 1000).until(
                    EC.presence_of_element_located((By.ID, "captcha-form"))
                )

            driver.quit()
            return keywords
    except Exception as e:
        print("error get keywords", e)
        return [{"autocomplete": []}]
    finally:
        driver.quit()
